File: code/day10.R
differences <- function(jolts, rating) {
  jolts <- sort(jolts)
  d1 <- d3 <- 0
  while (rating < max(jolts)) {
    plaus <- seq(rating + 1, rating + 3)
    next_rating <- min(plaus[which(plaus %in% jolts)])
    if (next_rating - rating == 1) {
      d1 <- d1 + 1
    } else if (next_rating - rating == 3) {
      d3 <- d3 + 1
    }
    rating <- next_rating
  }
  d1 * (d3 + 1)
}

get <- function(x, name) {
  val <- x[[as.character(name)]]
  ifelse(is.null(val), 0, val)
}

ans <- list()
ans[["0"]] <- 1
for (j in sort(jolts)) {
  ans[[as.character(j)]] <- get(ans, j - 1) + get(ans, j - 2) + get(ans, j - 3)
}

jolts <- scan("data/day10")

# puzzle 1
differences(jolts, 0)

# puzzle 2
sprintf("%.0f", ans[[length(jolts) + 1]])

File: code/day10.jl
function differences(file)
    jolts = parse.(Int, readlines(file)) |> sort |> x -> pushfirst!(x, 0)
    diffs = [jolts[i + 1] - jolts[i] for i in 1:(length(jolts) - 1)]
    sum(diffs .== 1) * (sum(diffs .== 3) + 1)
end

function paths(file)
    jolts = parse.(Int, readlines(file)) |> sort
    ans = Dict{Int, Int}(0 => 1)
    for jolt in jolts
        ans[jolt] = get(ans, jolt - 1, 0) + get(ans, jolt - 2, 0) + get(ans, jolt - 3, 0)
    end
    get(ans, maximum(jolts), 0)
end
